spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/handler/predicate/PathRoutePredicateFactory.java

```diff
@@ -90,8 +91,8 @@ public ShortcutType shortcutType() {
 		return new GatewayPredicate() {
 			@Override
 			public boolean test(ServerWebExchange exchange) {
-				PathContainer path = parsePath(
-						exchange.getRequest().getURI().getRawPath());
+				PathContainer path = parsePath(StringUtils
```

Comment:
Should this be an option?

Reply:
Yes `getRawPath` can return `null` but `cleanPath` handles that case and just returns `null` if it is

Reply:
Sorry. What I meant is, shouldn't clean path be opt in?

Reply:
I dont think so, do you see harm in it always being enabled?  I guess it could change the behavior but I think of this more as a bug that the path matches as the moment so I see this as a bug fix and the way it should have worked to begin with.

Reply:
It's changing the behavior, so it must be an opt in, no? Can you give an example of where it was a bug that this fixes? It looks like that request would have be an invalid request. If a user wants to send such a strange request then they can turn on this option...

Reply:
Also this filter will run for every request...if you look at StringUtils.cleanPath()...it is potentially not cheap...which I think is another reason it should be opt in.

Reply:
For example if you made a request that looked like `/foo/../bar/hello`.  If the only route you had created was a path predicate that matched `/foo/**`, this would be routed, when it really shouldn't.

I am pretty sure WebFlux is using `cleanPath` in their code to do the same thing and they dont seem to be concerned about it.

Reply:
Good point. Let's see if we can get a comment from @rstoyanchev or @rwinch

Reply:
We [do not apply](https://github.com/spring-projects/spring-framework/blob/5836680490afca32ba6070bd524fafc89b6c2342/spring-web/src/main/java/org/springframework/http/server/reactive/AbstractServerHttpRequest.java#L81) `StringUtils#cleanPath` in WebFlux for request mapping. I will defer to Rob's opinion on whether that is more secure to do so or not but to my knowledge unless the path is normalized early enough so that other parts of the code (especially URL based security) cannot see the original path, then it is more correct to not modify it. In addition, cleaning the raw (encoded) path means that `StringUtils` will skip over encoded dots which then when decoded could have the same meaning and might affect some code paths.

Reply:
@rstoyanchev thanks for the insight.  I can't find where I saw WebFlux using this but I saw it at some point.  I guess it is not happening all the time though as you pointed out.  After discussing this with the team I think we will introduce a property so users can opt in to this functionality.

Reply:
To be more clear I'm not sure it should be done at all. Is there a good reason to support something that is known to invite misuse? 

> I can't find where I saw WebFlux

Possibly in serving static resource. That we do but it's only because we have to resolve to an actual file on the file system and there are a lot of extra checks that we perform to ensure the final resource serviced is not outside the designated root from which it should be served.

Reply:
> To be more clear I'm not sure it should be done at all. Is there a good reason to support something that is known to invite misuse?

I am not sure what you mean by inviting misuse?  Why would cleaning the path invite misuse?

Reply:
IMO because nobody should be requesting with a path that requires cleaning up; should we even support requests like /foo/../bar/hello?

Reply:
@ryanjbaxter I tend to assume this is common knowledge, because we have dealt with many security issues, but I do forget not everyone knows the dangers lurking, which is what makes it all the more dangerous. As a starting point I suggest the information under Spring Security's own [DefaultHttpFirewall](https://docs.spring.io/spring-security/site/docs/current/api/org/springframework/security/web/firewall/DefaultHttpFirewall.html) and [StrictHttpFirewall](https://docs.spring.io/spring-security/site/docs/current/api/org/springframework/security/web/firewall/StrictHttpFirewall.html). You can search for "traversal" under https://pivotal.io/security which is one common misuse.

Reply:
@rstoyanchev I understand the implications I guess I just didn't understand what you mean by invite misuse.  Gateway users could add Spring Security and use the `DefaultHttpFirewall` or `StrictHttpFirewall` in their app if they wanted to.

This issue is really about how we the Path predicate decides whether a request matches or not.

Reply:
@tony-clarke-amdocs it is hard to say, as soon as we decide not to, someone might come back and say they were relying on that functionality.  Whether that is the correct thing to do or not is a different question.  The fact of the matter is that these requests are valid so IMO there is no reason why we should stop someone from supporting them if they choose to for whatever reason.